server/auth/azure.js

```diff
@@ -0,0 +1,88 @@
+// @flow
+import passport from "@outlinewiki/koa-passport";
+import jwt from "jsonwebtoken";
+import Router from "koa-router";
+import { capitalize } from "lodash";
+import { Strategy as AzureStrategy } from "passport-azure-ad-oauth2";
+import accountProvisioner from "../commands/accountProvisioner";
+import env from "../env";
+import auth from "../middlewares/authentication";
+import passportMiddleware from "../middlewares/passport";
+import { StateStore } from "../utils/passport";
+
+const router = new Router();
+const providerName = "azure";
+const AZURE_CLIENT_ID = process.env.AZURE_CLIENT_ID;
+const AZURE_CLIENT_SECRET = process.env.AZURE_CLIENT_SECRET;
+
+const scopes = [];
+
+export const config = {
+  name: "Microsoft",
+  enabled: !!AZURE_CLIENT_ID,
+};
+
+if (AZURE_CLIENT_ID) {
+  const strategy = new AzureStrategy(
+    {
+      clientID: AZURE_CLIENT_ID,
+      clientSecret: AZURE_CLIENT_SECRET,
+      callbackURL: `${env.URL}/auth/azure.callback`,
+      useCommonEndpoint: true,
+      passReqToCallback: true,
+      store: new StateStore(),
+      scope: scopes,
+    },
+    async function (req, accessToken, refreshToken, params, _, done) {
+      try {
+        // see docs for what the fields in profile represent here:
+        // https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens
+        const profile = jwt.decode(params.id_token);
+
+        const domain = profile.email.split("@")[1];
```

Comment:
Keep in mind that `profile.email` is not always present. In my company profile I can only see `profile.upn`, which contains my email address. In my personal account (which is a mix of password and github authentication) I have a mess.
I guess a request to `https://graph.microsoft.com/v1.0/me` would be better, since all data seem to be normalized, there